aspnetcore/blazor/Security-guidance.md

```diff
@@ -0,0 +1,254 @@
+
+### CPU exhaustion
+CPU exhaustion can happen when we allow a client to force the server to do intensive CPU work. 
+
+For example, lets say we build an application that allows a client to enter a number and have the server print the fibonnacci number for that number.
+
+Assuming that the server has a naive implementation, the client can then input a really big number and have the server perform a lot of work to determine the fibonnacci sequence for that number.
+
+This example, although a bit unreal, ilustrates the point. The client just needs to introduce a big number and the server will spend way more time than the client calculating it.
+
+This is a valid concern for all public facing applications, but while in general applications requests and connections time out, in blazor server-side applications that is not the case.
+
+As an application developer we must evaluate what are we asking the server to do when we answer to an event on the client, and include any appropriate check before we start doing any work.
+
+### Memory exhaustion
+Memory exhaustion can happen when we allow a client to force the server to consume a lot of memory.
+
+For example, lets say we build a todo-list application and that we don't limit the number of items that can be added to a list or that we naively render all the items at the same time.
+
+A client can then add items to the list non-stop, forcing the server to render all those items and consuming a lot of memory in the process.
+
+As a developer, there are several aspects of this problem that we must be aware of:
+  * When we hold data in a component like in our `TodoList` component, if we store the todo items in a `List<ToDoItem>` property or field, that list is using the server memory so if we allow that list to grow unbounded we run the risk of running out of memory. This causes not only the current session to end (crash) but all the sessions in that server instance at the time as the application receives an OutOfMemory exception.
+    * We recommend that you use some data structure with limits that you can configure to prevent this situation.
+  * When we force the UI to render all the todo items, the server is forced to hold on to additional memory for things that might not be visible in the screen at the time. (This assumes you don't render all the items and have some sort of paging)
+    * We recommend that you only render the components/data that needs to be visible to the user and avoid holding on to any other data in memory, specially if this data can grow based on user input like our `TodoList`. In this particular case you could achieve this by implementing some sort of _virtual-list_ that only exposes a window to the subset of elements at the current position/page for the component to render and keeps the rest either in secondary storage (ideal) or in memory (less ideal).
```

Comment:
Although this advice is good, talking about virtualization makes it sound so complicated that most devs will shrug and think it's unrealistic to handle it. I'd recommend suggesting a more basic strategy first, for example:

```suggestion
    * We recommend that you place limits on the number of items that can be rendered in a list. For example, use paginated lists, or simply display only the first 100 or 1000 items, requiring the user to enter more specific search criteria to find items later in the sequence. In a more advanced case, you could implement lists or grids that support *virtualization*, meaning that they only render the subset of items currently visible to the user even though the scrollbar allows the user to scroll through the complete set. The items not currently visible may be held in secondary storage (ideal) or in memory (less ideal).
```